opentelemetry-sdk/src/logs/growable_array.rs

```diff
@@ -0,0 +1,369 @@
+        for i in 0..15 {
+            collection.push(i);
+        }
+        for i in 0..15 {
+            assert_eq!(collection.get(i), Some(&(i as i32)));
+        }
+    }
+
+    #[test]
+    fn test_len() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        assert_eq!(collection.len(), 15);
+    }
+
+    #[test]
+    fn test_into_iter() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        let mut iter = collection.into_iter();
+        for i in 0..15 {
+            assert_eq!(iter.next(), Some(i));
+        }
+        assert_eq!(iter.next(), None);
+    }
+
+    #[test]
+    fn test_ref_iter() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..15 {
+            collection.push(i);
+        }
+        let iter = collection.iter();
+        let mut count = 0;
+        for value in iter {
+            assert_eq!(*value, count);
+            count += 1;
+        }
+        assert_eq!(count, 15);
+    }
+
+    #[test]
+    fn test_key_value_pair_storage_growable_array() {
+        let mut collection = GrowableArray::<KeyValuePair>::new();
+
+        let key1 = Key::from("key1");
+        let value1 = AnyValue::String("value1".into());
+        let key2 = Key::from("key2");
+        let value2 = AnyValue::Int(42);
+
+        collection.push(KeyValuePair(key1.clone(), value1.clone()));
+        collection.push(KeyValuePair(key2.clone(), value2.clone()));
+
+        assert_eq!(
+            collection.get(0).map(|kv| (&kv.0, &kv.1)),
+            Some((&key1, &value1))
+        );
+        assert_eq!(
+            collection.get(1).map(|kv| (&kv.0, &kv.1)),
+            Some((&key2, &value2))
+        );
+        assert_eq!(collection.len(), 2);
+
+        // Test iterating over the key-value pairs
+        let mut iter = collection.into_iter();
+        assert_eq!(iter.next(), Some(KeyValuePair(key1, value1)));
+        assert_eq!(iter.next(), Some(KeyValuePair(key2, value2)));
+        assert_eq!(iter.next(), None);
+    }
+
+    #[test]
+    fn test_empty_attributes() {
+        let collection = GrowableArray::<KeyValuePair>::new();
+        assert_eq!(collection.len(), 0);
+        assert_eq!(collection.get(0), None);
+
+        let mut iter = collection.into_iter();
+        assert_eq!(iter.next(), None);
+    }
+
+    #[test]
+    fn test_less_than_max_stack_capacity() {
+        let mut collection = GrowableArray::<i32>::new();
+        for i in 0..DEFAULT_MAX_STACK_CAPACITY - 1 {
+            collection.push(i as i32);
+        }
+        assert_eq!(collection.len(), DEFAULT_MAX_STACK_CAPACITY - 1);
+
+        for i in 0..DEFAULT_MAX_STACK_CAPACITY - 1 {
+            assert_eq!(collection.get(i), Some(&(i as i32)));
+        }
+        assert_eq!(collection.get(DEFAULT_MAX_STACK_CAPACITY), None);
```

Comment:
The next index after inserting elements in the collection here should be `DEFAULT_MAX_STACK_CAPACITY - 1`?

```suggestion
        assert_eq!(collection.get(DEFAULT_MAX_STACK_CAPACITY - 1), None);
```

Reply:
For this test, `DEFAULT_MAX_STACK_CAPACITY - 1` is the first out of bounds index. Shouldn't we test that since we are adding `DEFAULT_MAX_STACK_CAPACITY - 2` elements to the collection?

Reply:
ok I got what you mean. Have also added for `DEFAULT_MAX_STACK_CAPACITY - 1`. `DEFAULT_MAX_STACK_CAPACITY` was to ensure to test past the max size.